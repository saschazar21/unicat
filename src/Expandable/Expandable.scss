@import '../__styles__/variables';

.wrapper {
  position: relative;
  border-radius: borderRadius();
  border: border();
  box-shadow: boxShadow();
  background-color: color(white, false, $systemColors);
  padding: 0 spacing();
  margin-bottom: spacing(xs);
}

.activator {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: calc(100% + #{spacing() * 2});
  padding: spacing();
  margin: 0 -#{spacing()};
}

.icon {
  transition: transition(transform);
  min-height: fontSize(h4);
  min-width: fontSize(h4);
  transform: rotateZ(90deg);
  
  .open & {
    transform: rotateZ(-90deg);
  }
}

.title {
  margin-right: spacing();
  margin-bottom: 0;
}

.content {
  padding: spacing() 0;
  overflow: hidden;
}
