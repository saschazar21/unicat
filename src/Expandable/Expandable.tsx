import React, {
  useState,
  ReactNode,
  SyntheticEvent,
  KeyboardEvent,
} from 'react';
import classnames from 'classnames';
import { ChevronIcon } from '@saschazar/unicat-icons';
import AnimateHeight from 'react-animate-height';

import Heading from '../Heading';
import { preventDefault } from '../__tools__/helpers';

import styles from './Expandable.scss';

export const ANIMATION_DEFAULT_DURATION = 300;

export interface ExpandableProps {
  /* mandatory child */
  children: ReactNode;
  /* custom className */
  className?: string;
  /* is open by default? */
  open?: boolean;
  /* the title, visible at all times */
  title: string;
}

export default function expandable(props: ExpandableProps) {
  const { children, className: customClassname, open: isOpen, title } = props;
  const [open, setOpen] = useState(!!isOpen);

  const className = classnames(
    { [styles.open]: open },
    styles.activator,
    customClassname
  );

  const handleClick = (event: SyntheticEvent) => {
    preventDefault(event);
    setOpen(o => !o);
  };

  const handleKeyUp = (event: KeyboardEvent) => {
    preventDefault(event);
    if (event.key === 'ArrowUp' && open) {
      handleClick(event);
    }
    if (event.key === 'ArrowDown' && !open) {
      handleClick(event);
    }
    if (event.key === ' ') {
      handleClick(event);
    }
  };

  return (
    <aside className={styles.wrapper} role="region">
      <button
        className={className}
        onClick={handleClick}
        onKeyUp={handleKeyUp}
        aria-expanded={open}
        aria-label={title}
      >
        <Heading level="h5" className={styles.title}>
          {title}
        </Heading>
        <ChevronIcon
          focusable="false"
          aria-hidden="true"
          className={styles.icon}
        />
      </button>
      <AnimateHeight
        duration={ANIMATION_DEFAULT_DURATION}
        height={open ? 'auto' : 0}
      >
        <section className={styles.content}>{children}</section>
      </AnimateHeight>
    </aside>
  );
}
