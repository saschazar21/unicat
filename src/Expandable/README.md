```jsx
<Expandable title="Lorem ipsum">
  <p>
    Aut aut odit autem quam natus. Omnis minima unde a dolorem. Voluptatem
    ratione explicabo ab quaerat aliquid. Modi ut reiciendis deleniti qui error
    cumque. Libero et eligendi dicta totam consequatur sit fuga fugiat. Porro
    fugit nostrum quae iure. Nostrum aspernatur omnis sequi. Eum in qui suscipit
    consectetur. Similique nemo autem consequuntur perspiciatis earum voluptatem
    qui. Est rerum ipsum et neque rerum. Qui et porro dicta laborum ipsum.
    Officiis eveniet ad qui quaerat nam sunt. Ut autem est sint possimus beatae
    in dolore. Voluptas laboriosam quasi id officiis quia ratione. Numquam quis
    modi architecto et quos. Velit qui voluptatem beatae quia saepe corrupti
    iste. Sed reiciendis totam aperiam est. Voluptatem amet qui distinctio aut.
    Dolores corrupti quis illo. Eos magnam consequatur est sunt eveniet est in.
    Adipisci voluptas reprehenderit error sequi. Velit nam non iusto odio veniam
    voluptatem.
  </p>
</Expandable>
```
