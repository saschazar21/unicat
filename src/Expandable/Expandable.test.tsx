import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Expandable from './index';

describe('Expandable', () => {
  it('renders', () => {
    const wrapper = shallow(
      <Expandable title="Test">
        <p>Test Body</p>
      </Expandable>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
