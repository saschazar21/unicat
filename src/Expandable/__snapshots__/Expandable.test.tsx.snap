// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expandable has the .open class 1`] = `
<aside
  className="wrapper"
  role="region"
>
  <button
    aria-expanded={true}
    aria-label="Test"
    className="open activator"
    onClick={[Function]}
    onKeyUp={[Function]}
  >
    <Element
      SEO={false}
      className="title"
      level="h5"
    >
      Test
    </Element>
    <ChevronIcon
      aria-hidden="true"
      className="icon"
      focusable="false"
    />
  </button>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={300}
    easing="ease"
    height="auto"
    style={Object {}}
  >
    <section
      className="content"
    >
      <p>
        Test Body
      </p>
    </section>
  </AnimateHeight>
</aside>
`;

exports[`Expandable renders 1`] = `
<aside
  className="wrapper"
  role="region"
>
  <button
    aria-expanded={false}
    aria-label="Test"
    className="activator"
    onClick={[Function]}
    onKeyUp={[Function]}
  >
    <Element
      SEO={false}
      className="title"
      level="h5"
    >
      Test
    </Element>
    <ChevronIcon
      aria-hidden="true"
      className="icon"
      focusable="false"
    />
  </button>
  <AnimateHeight
    animateOpacity={false}
    animationStateClasses={
      Object {
        "animating": "rah-animating",
        "animatingDown": "rah-animating--down",
        "animatingToHeightAuto": "rah-animating--to-height-auto",
        "animatingToHeightSpecific": "rah-animating--to-height-specific",
        "animatingToHeightZero": "rah-animating--to-height-zero",
        "animatingUp": "rah-animating--up",
        "static": "rah-static",
        "staticHeightAuto": "rah-static--height-auto",
        "staticHeightSpecific": "rah-static--height-specific",
        "staticHeightZero": "rah-static--height-zero",
      }
    }
    applyInlineTransitions={true}
    delay={0}
    duration={300}
    easing="ease"
    height={0}
    style={Object {}}
  >
    <section
      className="content"
    >
      <p>
        Test Body
      </p>
    </section>
  </AnimateHeight>
</aside>
`;
